feat(dashboards): Multi-type time series visualizations

gggritso opened the pull request:
This PR allows plotting a line and a bar on the same plot. For now, they have to be of the same type, one thing at a time! This is the first part of https://github.com/getsentry/team-performance/issues/55

**e.g.,**
![Screenshot 2025-03-04 at 4 04 02 PM](https://github.com/user-attachments/assets/6ffb33ae-b374-410c-ad3d-b67e8f10d84d)

## Changes

This PR is basically just one change to the API of `TimeSeriesVisualizationWidget`. It _used_ to have a `timeSeries` prop, and it accepted an array of `TimeSeries` objects. `TimeSeries` is a data format, it's basically what comes back from the server. The visualization would convert those `TimeSeries` objects to `Plottable` objects under-the-hood, and then pass those off to ECharts. This approach stunk for a few reasons:

1. `TimeSeries` mixes data concerns and presentation concerns. Why is `color` part of `TimeSeries`? Data doesn't have a color!
2. It was impossible to plot lines and bars on top of each other. We want to do that!
2. It was impossible to set custom stacking and data delay per series. We want to do that, too!

This was the old API:

```jsx

const timeSeries = {
  field: 'eps()',
  color: 'red',
  data: [...

<TimeSeriesWidgetVisualization
  visualizationType="line"
  dataCompletenessDelay="60"
  timeSeries={[timeSeries]}
...
```

The new API is much more flexible. Those `Plottable` objects that we used under the hood are now the public API. They separate data and presentation, and can be configured per-series! Also, now that the component accepts `Plottable` objects, it can take advantage of getters like `isEmpty` and `needsColor` to handle different cases.

Here's the new API:

```jsx
<TimeSeriesWidgetVisualization
  plottables={[new Line(timeSeries, {color: 'red', delay: 60})
  ...
```

The `dataVisualizationDelay` and `visualizationType` props are not needed, they are set per-series.

This PR updates the component, updates all current usages of `TimeSeriesWidgetVisualization`, and adds new documentation to explain how to use the new props.

## Outstanding Questions

I'm still iterating on this. There are two things I'm interested in _maybe_ changing.

1. The construction pattern `new Line` doesn't feel great to me. I _think_ I'd rather have some kind of namespaced object, and do `Plottables.Line`, but I'm not confident that's better
2. The prop name `plottables` has grown on me while making this, and it nicely accommodates future incoming work (releases, rectangular markers, etc.) but it's not super intuitive. Maybe another name is better?

gggritso committed:
Multi-type, LFG

gggritso commented on static/app/views/dashboards/widgets/timeSeriesWidget/plottables/bars.tsx:
```diff
@@ -34,8 +34,7 @@ export class Bars extends ContinuousTimeSeries<BarsConfig> implements Plottable
         name: timeSeries.field,
-        color,
-        stack: config.stack ?? GLOBAL_STACK_NAME,
+        stack: config.stack,
```
Stack used to be required, because bar charts had all-or-nothing stacking. Now, stacking can be fully configurable, so we don't have a use for this fallback.

gggritso commented on static/app/views/explore/multiQueryMode/queryVisualizations/chart.tsx:
```diff
@@ -295,9 +309,13 @@ export function MultiQueryModeChart({
-          visualizationType={visualizationType}
-          timeSeries={chartInfo.data}
+          plottables={chartInfo.data.map(timeSeries => {
+            return new DataPlottableConstructor(timeSeries, {
+              delay: INGESTION_DELAY,
+              color: isTimeSeriesOther(timeSeries) ? theme.chartOther : undefined,
+              stack: 'all',
```
Handling of "other" series is moving out of `TimeSeriesWidgetVisualization` for now. I might bring it back later, but this is a ... leaky thing to try and look at.

gggritso committed:
Add mock data

gggritso committed:
Remove color from `TimeSeries`

gggritso committed:
More mock meta

gggritso committed:
Whoops

billyvg commented on static/app/views/dashboards/widgets/timeSeriesWidget/timeSeriesWidgetVisualization.tsx:
```diff
@@ -140,15 +118,14 @@ export function TimeSeriesWidgetVisualization(props: TimeSeriesWidgetVisualizati
 
-  // TODO: The `meta.fields` property should be typed as
-  // Record<string, AggregationOutputType | null>, which is the reality
+  // Determine out chart Y axis type and units
```
typo?

```suggestion
  // Determine our chart Y axis type and units
```

gggritso replied:
Yes, thank you!

billyvg reviewed (commented):
what's the difference between a plottable and a series?

gggritso commented:
@billyvg `Plottable` is _more_, basically. The major differences:

- `Series` does not keep track of the original data. It loses the `meta`, so it's not aware of data types, or units. `Plottable` has the original data, so `TimeSeriesWidgetVisualization` can check the units, and coordinate the Y axis and tooltips as necessary (by falling back, converting units, etc.)
- `Plottable` has some extra properties that `Series` isn't aware of. For now, that's only `delay`, which is a Sentry concept, and represents our ingestion delay, so we can mark incomplete buckets. A few others might show up when we start plotting samples
- `Series` mixes data and presentation. This isn't a huge issue, but doesn't feel good that `color` is jammed alongside the data
- `Plottable` is converted to N `Series` at render time. e.g., we plot dotted lines to show incomplete buckets by splitting the data into two, and creating a `Series` for each one. When we plot samples, each sample is a `Series`. Releases is similar. This is actually the _main_ reason, TBH

There are other differences, too

gggritso committed:
Fix typo

Co-authored-by: Billy Vong <billyvg@users.noreply.github.com>

gggritso committed:
Merge branch 'master' into feat/dashboards/time-series-multi-type

gggritso committed:
Resolve bad merge

gggritso committed:
Merge branch 'master' into feat/dashboards/time-series-multi-type

gggritso committed:
Add plottable start and end

This is useful for computing the boundaries of a plottable

gggritso committed:
Calculate the boundaries using Plottable

narsaynorath commented on static/app/views/insights/common/components/insightsTimeSeriesWidget.tsx:
```diff
@@ -48,18 +51,22 @@ export function InsightsTimeSeriesWidget(props: InsightsTimeSeriesWidgetProps) {
-    timeSeries: (props.series.filter(Boolean) ?? [])?.map(serie => {
+    plottables: (props.series.filter(Boolean) ?? [])?.map(serie => {
       const timeSeries = convertSeriesToTimeseries(serie);
+      const PlottableDataConstructor =
+        props.visualizationType === 'line'
+          ? Line
+          : props.visualizationType === 'area'
+            ? Area
+            : Bars;
```
Minor comment because I see this is only used in two places, but maybe we would want to extract this out into a separate util/factory?

gggritso replied:
Yes, I agree! I'll do that in a follow-up, when I integrate the `PlotType` type

narsaynorath reviewed (commented):
Some small comments, but I actually noticed a regression in the color palette selection. I'll poke around and see if I can spot anything but here are some of the charts in your story where the colours seem off (note the colors in the tooltip vs the legend under the header)

<img width="473" alt="Screenshot 2025-03-07 at 2 01 34 PM" src="https://github.com/user-attachments/assets/d3778578-adc2-48aa-865b-73b4a6ef5549" />
<img width="496" alt="Screenshot 2025-03-07 at 2 01 23 PM" src="https://github.com/user-attachments/assets/f8bb8a1c-c47b-49d9-b74f-5b25adec9c52" />

gggritso committed:
Merge branch 'master' into feat/dashboards/time-series-multi-type

gggritso committed:
Assign color to Bars series

The missing colour causes a mismatch from the series legend to the
actual series colour!

gggritso commented on static/app/views/dashboards/widgets/timeSeriesWidget/plottables/plottable.tsx:
```diff
@@ -19,6 +19,10 @@ export interface Plottable {
    * If the plottable is based on data, the unit. Otherwise, null
    */
   dataUnit: DurationUnit | SizeUnit | RateUnit | null;
+  /**
+   * Start timestamp of the plottable, if applicable
+   */
+  end: string | null;
```
You're right, it's for calculating the boundaries in the chart. I want `timeSeries` to be essentially private, I think it's a bad idea for downstream code to access the raw data stored in the plottable. So, I've had to add some helper methods to extract data bout the time series

gggritso committed:
Explain features

gggritso commented on static/app/views/dashboards/widgets/timeSeriesWidget/timeSeriesWidgetVisualization.stories.tsx:
```diff
@@ -57,10 +60,15 @@ export default storyBook('TimeSeriesWidgetVisualization', (story, APIReference)
             <li>displaying incomplete ingestion buckets</li>
+            <li>multi-type plotting</li>
+            <li>
+              support for <code>null</code> values
```
That is correct! I added a slightly more specific explanation, and improved the other bullet points too while I was at it

gggritso committed:
Add note about bar charts

gggritso commented on static/app/views/dashboards/widgets/timeSeriesWidget/timeSeriesWidgetVisualization.stories.tsx:
```diff
@@ -72,37 +80,90 @@ export default storyBook('TimeSeriesWidgetVisualization', (story, APIReference)
+          </li>
+          <li>
+            Bar charts are to your discretion, it's most an aesthetic choice. Generally,
+            bars communicate discrete buckets, and lines communicate continuous data. If
```
Oh that's interesting! I'll add that, thanks

gggritso commented on static/app/views/insights/common/components/insightsTimeSeriesWidget.tsx:
```diff
@@ -32,7 +35,7 @@ export interface InsightsTimeSeriesWidgetProps {
   series: DiscoverSeries[];
   title: string;
-  visualizationType: TimeSeriesWidgetVisualizationProps['visualizationType'];
+  visualizationType: 'line' | 'area' | 'bar';
```
There's actually already a type for this, called `PlotType`! I'm aiming to use that everywhere, but haven't had a chance yet. I'll need to put it somewhere central, and update a bit of code to make this it's actually in use

gggritso committed:
Extract ingestion delay setting

gggritso commented:
@narsaynorath right on, thanks! I've addressed the line comments (I think), and I fixed the colour bug. It was caused by a missing `color` property on the `Bars` constructor. I was only setting the item colour but not the _series_ colour, so ECharts was using a fallback, which was mismatched from the actual series 👍🏻

narsaynorath reviewed (approved):
:tada:
